src/inmanta/agent/agent.py

```diff
@@ -1332,63 +1597,93 @@ async def on_disconnect(self) -> None:
         """
-        Ensure that the code for the given environment and version is loaded.
-        Return a list of all the ``resource_types`` that it failed to load.
+        Get the list of installation specifications (i.e. pip congif, python package dependencies, Inmanta modules sources)
+        required to deploy a given version for the provided resource types.
         """
-        failed_to_load: set[str] = set()
         if self._loader is None:
-            return failed_to_load
+            return []
 
         # store it outside the loop, but only load when required
         pip_config: Optional[PipConfig] = None
 
-        for rt in set(resource_types):
+        resource_install_specs: list[ResourceInstallSpec] = []
+        for resource_type in set(resource_types):
+
+            if self._last_loaded[resource_type] == version:
+                LOGGER.debug("Code already present for %s version=%d", resource_type, version)
+                continue
             # only one logical thread can load a particular resource type at any time
-            async with self._resource_loader_lock.get(rt):
+            # TODO deal with this lock  -> not sure how
+            async with self._resource_loader_lock.get(resource_type):
+                result: protocol.Result = await self._client.get_source_code(environment, version, resource_type)
+                if result.code == 200 and result.result is not None:
+                    sync_client = SyncClient(client=self._client, ioloop=self._io_loop)
+                    LOGGER.debug("Installing handler %s version=%d", resource_type, version)
+                    requirements: set[str] = set()
+                    sources: list["ModuleSource"] = []
+                    # Encapsulate source code details in ``ModuleSource`` objects
+                    for source in result.result["data"]:
+                        sources.append(
+                            ModuleSource(
+                                name=source["module_name"],
+                                is_byte_code=source["is_byte_code"],
+                                hash_value=source["hash"],
+                                _client=sync_client,
+                            )
+                        )
+                        requirements.update(source["requirements"])
+
+                    if pip_config is None:
+                        # TODO: what if this fails?
+                        pip_config = await self._get_pip_config(environment, version)
+
+                    resource_install_specs.append(
+                        ResourceInstallSpec(resource_type, version, pip_config, list(requirements), sources)
+                    )
+                else:
+                    # TODO what if server call fails ?
+                    pass
+
+        return resource_install_specs
+
+    async def ensure_code(self, code: Sequence[ResourceInstallSpec]) -> Set[str]:
```

Comment:
All of the source code handling is not quite in the right place yet, but I think we can gloss over that for now.